File: backend/app/schemas/import_export.py
from __future__ import annotations

from datetime import datetime
from typing import List, Literal

from pydantic import BaseModel, Field


class MiyaDeckMeta(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=1000)
    is_public: bool = False
    tags: list[str] | None = None


class MiyaCard(BaseModel):
    front_content: str = Field(..., min_length=1, max_length=5000)
    back_content: str = Field(..., min_length=1, max_length=5000)
    tags: list[str] | None = None
    hint: str | None = Field(default=None, max_length=2000)


class MiyaDeckPayload(BaseModel):
    deck: MiyaDeckMeta
    cards: list[MiyaCard]


class ImportCardPreview(BaseModel):
    temp_id: int
    front_content: str
    back_content: str
    tags: list[str] | None = None
    hint: str | None = None
    is_duplicate: bool = False
    selected: bool = True


class ImportDeckPreview(BaseModel):
    source_name: str
    deck: MiyaDeckMeta
    cards: list[ImportCardPreview]
    total_cards: int
    duplicate_cards: int


class ImportPreviewResponse(BaseModel):
    items: list[ImportDeckPreview]
    total_decks: int
    total_cards: int
    total_duplicates: int


class ImportApplyDeck(BaseModel):
    deck: MiyaDeckMeta
    cards: list[ImportCardPreview]
    conflict_mode: Literal["create_new", "merge_into_existing"] = "create_new"
    existing_deck_id: int | None = None


class ImportApplyRequest(BaseModel):
    decks: list[ImportApplyDeck]


class ImportApplyResult(BaseModel):
    deck_id: int
    imported_cards: int
    skipped_duplicates: int


class ImportApplyResponse(BaseModel):
    results: list[ImportApplyResult]


class ImportExportHistoryItem(BaseModel):
    id: int
    deck_id: int | None
    action: Literal["import", "export"]
    format: str
    total_cards: int | None
    details: str | None
    created_at: datetime


class ImportExportHistoryResponse(BaseModel):
    items: List[ImportExportHistoryItem]

File: frontend/src/services/importExport.ts
import { api } from "./api";

export interface MiyaDeckMeta {
  title: string;
  description?: string | null;
  is_public?: boolean;
  tags?: string[] | null;
}

export interface MiyaCard {
  front_content: string;
  back_content: string;
  tags?: string[] | null;
  hint?: string | null;
}

export interface ImportCardPreview {
  temp_id: number;
  front_content: string;
  back_content: string;
  tags?: string[] | null;
  hint?: string | null;
  is_duplicate: boolean;
  selected: boolean;
}

export interface ImportDeckPreview {
  source_name: string;
  deck: MiyaDeckMeta;
  cards: ImportCardPreview[];
  total_cards: number;
  duplicate_cards: number;
}

export interface ImportPreviewResponse {
  items: ImportDeckPreview[];
  total_decks: number;
  total_cards: number;
  total_duplicates: number;
}

export interface ImportApplyDeck {
  deck: MiyaDeckMeta;
  cards: ImportCardPreview[];
  conflict_mode: "create_new" | "merge_into_existing";
  existing_deck_id?: number | null;
}

export interface ImportApplyRequest {
  decks: ImportApplyDeck[];
}

export interface ImportApplyResult {
  deck_id: number;
  imported_cards: number;
  skipped_duplicates: number;
}

export interface ImportApplyResponse {
  results: ImportApplyResult[];
}

export interface ImportExportHistoryItem {
  id: number;
  deck_id: number | null;
  action: "import" | "export";
  format: string;
  total_cards: number | null;
  details: string | null;
  created_at: string;
}

export interface ImportExportHistoryResponse {
  items: ImportExportHistoryItem[];
}

export const importExportService = {
  async preview(files: File[], jsonText?: string): Promise<ImportPreviewResponse> {
    const form = new FormData();
    for (const file of files) {
      form.append("files", file);
    }
    if (jsonText && jsonText.trim().length > 0) {
      form.append("json_text", jsonText);
    }
    const response = await api.post<ImportPreviewResponse>("/decks/import-export/import/preview", form, {
      headers: { "Content-Type": "multipart/form-data" },
    });
    return response.data;
  },

  async apply(request: ImportApplyRequest): Promise<ImportApplyResponse> {
    const response = await api.post<ImportApplyResponse>("/decks/import-export/import/apply", request);
    return response.data;
  },

  async download(deckId: number, format: "json" | "md" | "pdf" | "apkg"): Promise<Blob> {
    const response = await api.get(`/decks/import-export/export/${deckId}`, {
      params: { format },
      responseType: "blob",
    });
    return response.data;
  },

  async history(): Promise<ImportExportHistoryItem[]> {
    const response = await api.get<ImportExportHistoryResponse>("/decks/import-export/history");
    return response.data.items;
  },
};

